packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -140,14 +182,25 @@ export class TagManager {
       include.indexOf(this.resourceTypeName) === -1) {
       return false;
     }
-    if (this.tags[key]) {
-      if (this.tags[key].props.priority !== undefined) {
-        return priority >= this.tags[key].props.priority!;
+    if (this.tags.has(key)) {
```

Comment:
@rix0rrr - good feedback. 

Quick follow **is this a push to remove include/exclude by resource type**?

I think the crux of this issue is: 
> Now, I understand that doesn't fit in the current model really well, where we don't even know the TYPE of the L1 tags until we get to the renderer, so I won't make you change it.

What if we change that? We have a format for tag input that is agnostic to output. If I change the code generation to generate a tag type that is uniform (for input) and unique on output (to keep validation the same as it is today), then this issue disappears. If we look at the current tag input structures:

```ts
    tags?: object | cdk.Token;
// OR
    tags?: Array<cdk.CfnTag | cdk.Token> | cdk.Token;
// OR
    tags?: Array<CfnAutoScalingGroup.TagPropertyProperty | cdk.Token> | cdk.Token;
// We standardize to
    interface CdkTag { // not sure on this name yet
       key: string;
       value: string;
       // default 50 or whatever we decide from above
       priority?: number;
       // default true
       applyToLaunchedInstances?: boolean;
    }
    tags?: CdkTag[];
```
I think with our ability to treat tokens as strings and the ability to now add tags later via aspect we no longer need the `... | cdk.Token` definitions anywhere. Now @eladb might beat me up on ergonomics here. Yes, I agree simple key, value would be better. However, this keeps the L1 in a true declarative state. The rule I'm clearly breaking here is, this is the first input property that doesn't directly match CloudFormation. If somebody was writing a generator to turn templates into CDK files this would now require special code. I personally think the standardization is a bigger win, and later if we do a declarative CDK at the L2 level we have a consistent definition of tag.

Then in the constructor to `TagManager` we can pass in initial tags and they are effectively applied immediately because these tags do not propagate like an aspect. Now we roll in your changes from above and I think this is a much cleaner code solution where `merge` isn't really necessary. If it is necessary we can create the `merge` on a tag class that does this algebra as you describe.

Reply:
Counter-proposal: we keep the L1 type definitions as they are (I do think there's value in sticking closely to the CFN model, if only for predictability and explainability), AND we define the one uber-type of tag, and we make the `TagFormatter` be able to convert between them:

```ts
interface ITagFormatter {
  parseTags(l1tags: any): CdkUberTag[];
  formatTags(tags: CdkUberTag[]): any;
}
```

So in fact, the tag *formatter* becomes a tag *converter*, back and forth from our unified L2 type to the various L1 types.

I *would* type the `any` more strictly there, but doing so would require generics and I think we might run afoul of JSII. It's going to require unnatural gymnastics to keep the definition of `ITagFormatter` out of reach of the JSII parser; it's probably simpler to type as `any` and add some runtime checks.

And then we can still use `TagFormatter/Converter` in the constructor to get the initial set of tags applied and we don't need to do any merging.